src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.CompilationTracker.cs

```diff
@@ -856,24 +852,17 @@ private Compilation CreateEmptyCompilation()
             public Task<bool> HasSuccessfullyLoadedAsync(
                 SolutionCompilationState compilationState, CancellationToken cancellationToken)
             {
-                var state = this.ReadState();
-
-                if (state.HasSuccessfullyLoaded.HasValue)
-                {
-                    return state.HasSuccessfullyLoaded.Value ? SpecializedTasks.True : SpecializedTasks.False;
-                }
-                else
-                {
-                    return HasSuccessfullyLoadedSlowAsync(compilationState, cancellationToken);
-                }
+                return this.ReadState() is FinalState finalState
+                    ? finalState.HasSuccessfullyLoaded ? SpecializedTasks.True : SpecializedTasks.False
+                    : HasSuccessfullyLoadedSlowAsync(compilationState, cancellationToken);
             }
 
             private async Task<bool> HasSuccessfullyLoadedSlowAsync(
                 SolutionCompilationState compilationState, CancellationToken cancellationToken)
```

Comment:
As a potential cleanup: the ...LoadedSlowAsync() might have been entirely to avoid extra task allocations for the fast-path of everything was already created; this might be a nice place to use ValueTask now.